### PR TITLE
add geometry column to Shapefile.Table

### PR DESCRIPTION
**Pull request by TheCedarPrince**
Currently a WIP. Issue linked here #48

### CODE REVIEW

**Comment by TheCedarPrince** [comment]
Hey @visr I began opening a PR - would you be alright with guiding me through this? I am a bit unclear on how to generate a Table schema from the Table _and_ Row together. Best thing I was thinking of was to somehow combine the Table and Row getters into one getdbf function.

File: src/table.jl [hunk 1]
```diff
@@ -46,6 +46,8 @@ end
 getshp(t::Table) = getfield(t, :shp)
 getdbf(t::Table) = getfield(t, :dbf)
 
+getshp(t::Row) = getfield(t, :geometry)
```
Comment (visr) [inline src/table.jl:49]: There is already the `shape` function below that does the same thing, so this isn't needed.

File: src/table.jl [hunk 2]
```diff
@@ -63,12 +65,14 @@ function Base.iterate(t::Table, st = 1)
...
 
 Base.getproperty(row::Row, name::Symbol) = getproperty(getfield(row, :record), name)
+Base.getproperty(row::Row, name::Symbol) = getproperty(getfield(row, :geometry), name)
 Base.getproperty(t::Table, name::Symbol) = getproperty(getdbf(t), name)
 
 Base.propertynames(row::Row) = propertynames(getfield(row, :record))
 Base.propertynames(t::Table) = propertynames(getdbf(t))
 
 Tables.schema(t::Table) = Tables.schema(getdbf(t))
+Tables.schema(t::Row) = Tables.schema(getdbf(t))
```
Comment (visr) [inline src/table.jl:75]: Is it normal to define a schema on a row? I thought this function was only meant for objects implementing the table interface, and Row by itself doesn't implement that.

Comment (TheCedarPrince) [reply src/table.jl:75]: Right! To be honest, I have not worked with `Table.schema` objects before so I wasn't sure what the best approach would be here. Do you happen to know of anything I could look at to help me in figuring this out? Thanks!

Comment (visr) [reply src/table.jl:75]: Here is the schema for the DBF part:
https://github.com/JuliaData/DBFTables.jl/blob/6b7ff043f9d5d9df0408006b7c612b9555a7315b/src/DBFTables.jl#L282-L288
So we should probably unpack that schema, and add geometry to the names, and the eltype of the shapes to the types.

File: src/table.jl (hunk 1 above)
Comment (TheCedarPrince) [reply src/table.jl:49]: Removed this

File: src/table.jl (hunk 2 above)
Comment (TheCedarPrince) [reply src/table.jl]: Ok, I think I was able to successfully do this @visr. Adding a new commit now.

**Comment by TheCedarPrince** [comment]
Hey @visr - I spent some time this morning prototyping a possible approach following our discussion on extending the table schema. What should I do next here? Thanks!

**Comment by TheCedarPrince** [comment]
Hey @visr - I actually went a bit further and got functionality working for getting `propertynames`, creating a `Tables.Schema` using a new function I created called `getall` and am now stuck at the `getproperty` function. How do things look so far? What are your thoughts? Thanks!

**Comment by visr** [comment]
```
julia> using Shapefile, Tables, DataFrames

julia> path = joinpath(dirname(pathof(Shapefile)),"..","test","shapelib_testcases","test.shp");

julia> table = Shapefile.Table(path)
Shapefile.Table{Union{Missing, Shapefile.Polygon}} with 3 rows and the following 11 columns:

Descriptio, TestInt, TestDouble, code, length, version, shapeType, MBR, zrange, mrange, shapes

julia> Tables.schema(table)
Tables.Schema:
 :Descriptio  Union{Missing, String}
 :TestInt     Union{Missing, Int64}
 :TestDouble  Union{Missing, Float64}
 :code        Int32
 :length      Int32
 :version     Int32
 :shapeType   Int32
 :MBR         Any
 :zrange      Any
 :mrange      Any
 :shapes      Union{Missing, Shapefile.Polygon}
```

The first 3 entries here are the columns from the DBF attribute table. Besides those 3, we only need to add 1 entry for the shapes, just like the current last entry. It would be my preference to call that column geometry and to put it as the first column, the other entries from `code` to `mrange`, are scalar header properties, and not part of the table data.

Perhaps leaving out those extra entries will help with the `getproperty` business as well. For `getproperty` on a Shapefile.Table, I'd expect something like:

```
function Base.getpropery(t::Table, name::Symbol)
    if name === :geometry
        shapes(t)
    else
        getproperty(getdbf(t), name)
    end
```

**Comment by TheCedarPrince** [comment]
Ah ok - I think we are getting closer to what I need to do. How would I add an entry to the DBF attribute table? The closest I got to this was creating a new attribute table as follows:

```julia
function getall(t::Table)
       names = []
       types = []
       map([getdbf(t), getshp(t)]) do values
       	   for prop in values |> propertynames
	   	push!(names, prop)
		push!(types, getproperty(values, prop) |> eltype) 
	   end
       end
       return Tables.Schema(tuple(names...), tuple(types...))
end
```

and then 

```julia
Tables.schema(t::Table) = getall(t)
```

Where the `getall` would return a proper schema which includes `:geometry` as the first column in the DBF. 

In the `getall`, instead of including every single attribute from the `getshp` command, I could just modify it to where it could only get the "shape" field, map it to a new symbol called ":geometry" and have it come first in the schema.

The `getproperty` idea sounds great and will pursue that as well.

Does it sound like I am getting better idea of things or not quite @visr ?

**Comment by visr** [comment]
Yep, that sounds right! I'd also see if you can easily avoid creating a new getall function, and just add it as a method for Tables.schema on Shapefile.Table.

**Comment by Sov-trotter** [comment]
Hey @TheCedarPrince. Here are some things you can look into
`for prop in values |> propertynames` works for dbf but not shp since that will end up fetching field names of the shapefile constituents 
`julia> Shapefile.getshp(t) |> propertynames
(:code, :length, :version, :shapeType, :MBR, :zrange, :mrange, :shapes)`
which we don't want. 
For this you can either hardcode a field/column name maybe `:Geometry`. Similarly for type you can use the `SHAPETYPE` constant defined earlier.
eg: `SHAPETYPE[getshp(t).shapeType])` Not sure if it's the best way to fetch the type though.

**Comment by TheCedarPrince** [comment]
@visr and everyone, this is ready for a review and hopefully for merging! Thanks!

**Review by visr — APPROVED** [review]
Thanks, this looks good to me. Since this changes the Shapefile.Table to include a geometry column, this should be a breaking change. I haven't yet checked if any of the dependencies would be hit by this though.

**Comment by martinmestre** [comment]
Hello @visr ! 
I would appreciate some help with Shapefile.Table(). I apply this function and then check with Tables.schema()
but I do not see the geometry. Am I missing something ? 
Thank you very much in advance.
Martín

**Comment by visr** [comment]
Hi Martín, is that using the latest version of this package, 0.7.1? And what kind of file?
If I use the example from the readme I see the geometry in the schema:

```julia
using Shapefile, Tables
path = joinpath(dirname(pathof(Shapefile)),"..","test","shapelib_testcases","test.shp")
table = Shapefile.Table(path)

julia> Tables.schema(table)
Tables.Schema:
 :geometry    Union{Missing, Shapefile.Polygon}
 :Descriptio  Union{Missing, String}
 :TestInt     Union{Missing, Int64}
 :TestDouble  Union{Missing, Float64}
```

If it's not working for you then it's probably best to create a new issue.

**Comment by martinmestre** [comment]
Thanks @visr, I had an older version. Now it works.

**Pull request merged**